In [3]:
import mysql.connector

In [4]:
import pandas as pd

In [5]:
mydb = mysql.connector.connect(
    host = 'localhost',
    user = 'root',
    passwd = 'Wonder1995!',
    database = 'retrowheels'
)

1.  Rumuskan single query untuk menampilkan jumlah total customer yang Anda layani, beserta jumlah
total kota & negara asal customer-customer Anda. 

In [6]:
mycursor = mydb.cursor()
query = 'SELECT COUNT(DISTINCT customernumber) as Customers, COUNT(DISTINCT city) as Cities, COUNT(DISTINCT country) as Countries FROM customers'
mycursor.execute(query)
result = mycursor.fetchall()
df = pd.DataFrame(result,columns = mycursor.column_names)
df

Customers  Cities  Countries
0        122      95         27

2. Rumuskan single query untuk menampilkan resources yang Anda miliki, mulai dari jumlah karyawan,
jumlah kantor & lokasi negaranya, jumlah barang yang dijual, total stok barang & jumlah vendor
yang menjadi partner Anda.

In [7]:
mycursor = mydb.cursor()
query = 'SELECT COUNT(distinct employeeNumber) as Employees,COUNT(distinct offices.officeCode) as Offices,COUNT(distinct country) as Country,COUNT(distinct productCode) as Products,SUM(distinct quantityinstock) as StockProducts,COUNT(distinct productVendor) as Vendors FROM employees,offices,products'
mycursor.execute(query)
result = mycursor.fetchall()
df = pd.DataFrame(result,columns = mycursor.column_names)
df

Employees  Offices  Country  Products StockProducts  Vendors
0         23        7        5       110        555131       13

3. Dari soal sebelumnya tercatat Anda memiliki 110 model die cast dengan total stok 555131 item. Jika
dikategorikan, produk yang Anda jual terbagi menjadi 7 product line die cast, yakni model mobil klasik,
mobil vintage, sepeda motor, pesawat terbang, kapal laut, kereta api serta truk & bus. Rumuskan single
query yang dapat menampilkan harga produk terendah & tertinggi untuk masing-masing kategori.

In [8]:
mycursor = mydb.cursor()
query = 'SELECT productLine,min(buyPrice),max(buyPrice) FROM products GROUP BY productLine'
mycursor.execute(query)
result = mycursor.fetchall()
df = pd.DataFrame(result,columns = mycursor.column_names)
df

productLine min(buyPrice) max(buyPrice)
0      Classic Cars         15.91        103.42
1       Motorcycles         24.14         91.02
2            Planes         29.34         77.27
3             Ships         33.30         82.34
4            Trains         26.72         67.56
5  Trucks and Buses         24.92         84.76
6      Vintage Cars         20.61         86.70

4. Rumuskan single query yang dapat menampilkan daftar 10 customer paling royal (paling banyak
mendatangkan uang bagi kita), yang total nominal transaksinya paling tinggi. Data yang ditampilkan
adalah nama customer, kota & negara asal, beserta total uang yang dihabiskan di produk kita

In [9]:
mycursor = mydb.cursor()
query = 'SELECT customers.customername,customers.city,customers.country,sum(amount) as total from payments LEFT JOIN customers on customers.customernumber = payments.customernumber group by payments.customernumber order by total DESC limit 10'
mycursor.execute(query)
result = mycursor.fetchall()
df = pd.DataFrame(result,columns = mycursor.column_names)
df

customername           city      country      total
0        Euro+ Shopping Channel         Madrid        Spain  715738.98
1  Mini Gifts Distributors Ltd.     San Rafael          USA  584188.24
2    Australian Collectors, Co.      Melbourne    Australia  180585.07
3            Muscle Machine Inc            NYC          USA  177913.95
4       Dragon Souveniers, Ltd.      Singapore    Singapore  156251.03
5    Down Under Souveniers, Inc     Auckland    New Zealand  154622.08
6                AV Stores, Co.     Manchester           UK  148410.09
7       Anna's Decorations, Ltd   North Sydney    Australia  137034.22
8      Corporate Gift Ideas Co.  San Francisco          USA  132340.78
9        Saveley & Henriot, Co.           Lyon       France  130305.35

5. Pada 2003-06-05, terdapat pembayaran masuk sebesar US$ 14571.44. Tampilkan data seputar
transaksi tersebut, mencakup nama customer yang melakukan pembayaran, nama produk yang dibeli,
jumlah tiap produk yang dibeli dan harga satuannya. Pastikan total harga yang dibeli sesuai dengan
data pembayaran masuk.

In [16]:
mycursor = mydb.cursor()
query = "SELECT customers.customerName, products.productname, quantityordered,priceEach FROM products RIGHT JOIN orderdetails ON products.productCode = orderdetails.productCode RIGHT JOIN orders ON orderdetails.ordernumber = orders.ordernumber RIGHT JOIN customers ON orders.customerNumber = customers.customerNumber LEFT JOIN payments ON customers.customernumber = payments.customernumber WHERE payments.paymentdate = '2003-06-05' AND orders.orderdate < '2003-06-05'"
mycursor.execute(query)
result = mycursor.fetchall()
df = pd.DataFrame(result,columns = mycursor.column_names)
df

customerName                     productname  quantityordered  \
0  Atelier graphique           1965 Aston Martin DB5               26   
1  Atelier graphique    1999 Indy 500 Monte Carlo SS               46   
2  Atelier graphique  1948 Porsche Type 356 Roadster               34   
3  Atelier graphique       1966 Shelby Cobra 427 S/C               50   

  priceEach  
0    120.71  
1    114.84  
2    117.26  
3     43.27